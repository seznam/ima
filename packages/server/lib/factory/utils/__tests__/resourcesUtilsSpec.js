import { afterEach, describe, expect, it, vi } from 'vitest';

const {
  renderStyles,
  prepareDefaultResources,
  renderScript,
} = require('../resourcesUtils');

const manifestMock = require('../../__mocks__/manifest.json');

vi.mock('fs', () => {
  const { toMockedInstance } = vi.requireActual('to-mock');
  const originalModule = vi.requireActual('fs');

  return {
    ...toMockedInstance(originalModule, {
      existsSync() {
        return true;
      },
      readFileSync(path) {
        if (path.endsWith('manifest.json')) {
          return JSON.stringify(manifestMock);
        }

        return 'runner#{source}';
      },
    }),
  };
});

describe('resourcesUtils', () => {
  describe('renderStyles', () => {
    it('should return empty string for invalid input', () => {
      expect(renderStyles()).toBe('');
      expect(renderStyles([])).toBe('');
      expect(renderStyles({})).toBe('');
      expect(renderStyles(null)).toBe('');
      expect(renderStyles('/static/app.css')).toBe('');
    });

    it('should return link stylesheet tags for string items', () => {
      expect(renderStyles(['/static/app.css'])).toMatchInlineSnapshot(
        `"<link as="style" href="/static/app.css" rel="preload" type="text/css" /><link rel="stylesheet" href="/static/app.css" />"`
      );

      expect(
        renderStyles([
          '/static/app1.css',
          '/static/app2.css',
          '/static/app3.css',
        ])
      ).toMatchInlineSnapshot(
        `"<link as="style" href="/static/app1.css" rel="preload" type="text/css" /><link as="style" href="/static/app2.css" rel="preload" type="text/css" /><link as="style" href="/static/app3.css" rel="preload" type="text/css" /><link rel="stylesheet" href="/static/app1.css" /><link rel="stylesheet" href="/static/app2.css" /><link rel="stylesheet" href="/static/app3.css" />"`
      );
    });

    it('should return link tag with custom attributes', () => {
      expect(
        renderStyles([['/static/app.css', { 'data-id': 'custom-id' }]])
      ).toMatchInlineSnapshot(
        `"<link as="style" href="/static/app.css" rel="preload" type="text/css" /><link href="/static/app.css" rel="stylesheet" data-id="custom-id" />"`
      );
    });

    it('should insert custom mini script for fallback sources', () => {
      expect(
        renderStyles([
          [
            '/static/app.css',
            { rel: 'stylesheet', fallback: '/static/fallback.css' },
          ],
        ])
      ).toMatchInlineSnapshot(
        `"<link as="style" href="/static/app.css" rel="preload" type="text/css" /><link href="/static/app.css" rel="stylesheet" onerror="this.onerror=null;this.href='/static/fallback.css';" />"`
      );
    });

    it('should disable link preload', () => {
      expect(
        renderStyles([
          [
            '/static/app.css',
            {
              rel: 'stylesheet',
              preload: false,
            },
          ],
        ])
      ).toMatchInlineSnapshot(
        `"<link href="/static/app.css" rel="stylesheet" />"`
      );
    });
  });

  describe('renderScript', () => {
    it('should return empty string for invalid input', () => {
      expect(renderScript('name')).toBe('');
      expect(renderScript(undefined, [])).toBe('');
      expect(renderScript('name', [])).toBe('');
      expect(renderScript({}, {})).toBe('');
      expect(renderScript('name', null)).toBe('');
    });

    it('should wrap script contents in script tag', () => {
      expect(
        renderScript('revival-settings', 'function () {}')
      ).toMatchSnapshot();
    });
  });

  describe('prepareDefaultResources', () => {
    afterEach(() => {
      process.env.IMA_PUBLIC_PATH = '';
    });

    it('should prepare default sources structure from provided manifest file', () => {
      expect(prepareDefaultResources(manifestMock, 'en')).toMatchSnapshot();
    });

    it('should add fallbacks when IMA_PUBLIC_PATH is defined', () => {
      process.env.IMA_PUBLIC_PATH = 'cdn://';

      expect(prepareDefaultResources(manifestMock, 'en')).toMatchSnapshot();
    });

    it('should add fallbacks when IMA_PUBLIC_PATH is defined, with proper custom config publicPath', () => {
      process.env.IMA_PUBLIC_PATH = 'cdn://';

      expect(
        prepareDefaultResources(
          {
            ...manifestMock,
            publicPath: '/pro/static/',
          },
          'en'
        )
      ).toMatchSnapshot();
    });

    it('should skip compilations without assets', () => {
      const sources = prepareDefaultResources(
        {
          ...manifestMock,
          assetsByCompiler: {
            ...manifestMock.assetsByCompiler,
            client: {},
          },
        },
        'en'
      );

      expect(sources).toMatchSnapshot();
    });
  });
});
