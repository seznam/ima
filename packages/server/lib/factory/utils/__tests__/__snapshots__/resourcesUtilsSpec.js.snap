// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`resourcesUtils > prepareDefaultResources > should add fallbacks when IMA_PUBLIC_PATH is defined 1`] = `
{
  "esScripts": [
    [
      "/static/js.es/app.client.2106a34c6b8bbad8.js",
      {
        "async": "",
        "crossorigin": "anonymous",
      },
    ],
    [
      "/static/js.es/locale/en.371127fdbfbe93d2.js",
      {
        "async": "",
        "crossorigin": "anonymous",
      },
    ],
    [
      "/static/js.es/vendors.a873907f25297544.js",
      {
        "async": "",
        "crossorigin": "anonymous",
      },
    ],
  ],
  "esStyles": [
    [
      "/static/css/app.d0ad44d05f82db5f.css",
      {
        "preload": true,
        "rel": "stylesheet",
      },
    ],
  ],
  "scripts": [
    [
      "/static/js/app.client.8dc14fd2c9e52eef.js",
      {
        "async": "",
        "crossorigin": "anonymous",
      },
    ],
    [
      "/static/js/locale/en.3c6a5e7a55bb2ab4.js",
      {
        "async": "",
        "crossorigin": "anonymous",
      },
    ],
    [
      "/static/js/vendors.0e456297851f0a3a.js",
      {
        "async": "",
        "crossorigin": "anonymous",
      },
    ],
  ],
  "styles": [],
}
`;

exports[`resourcesUtils > prepareDefaultResources > should add fallbacks when IMA_PUBLIC_PATH is defined, with proper custom config publicPath 1`] = `
{
  "esScripts": [
    [
      "cdn://static/js.es/app.client.2106a34c6b8bbad8.js",
      {
        "async": "",
        "crossorigin": "anonymous",
        "fallback": "/pro/static/static/js.es/app.client.2106a34c6b8bbad8.js",
      },
    ],
    [
      "cdn://static/js.es/locale/en.371127fdbfbe93d2.js",
      {
        "async": "",
        "crossorigin": "anonymous",
        "fallback": "/pro/static/static/js.es/locale/en.371127fdbfbe93d2.js",
      },
    ],
    [
      "cdn://static/js.es/vendors.a873907f25297544.js",
      {
        "async": "",
        "crossorigin": "anonymous",
        "fallback": "/pro/static/static/js.es/vendors.a873907f25297544.js",
      },
    ],
  ],
  "esStyles": [
    [
      "cdn://static/css/app.d0ad44d05f82db5f.css",
      {
        "fallback": "/pro/static/static/css/app.d0ad44d05f82db5f.css",
        "preload": true,
        "rel": "stylesheet",
      },
    ],
  ],
  "scripts": [
    [
      "cdn://static/js/app.client.8dc14fd2c9e52eef.js",
      {
        "async": "",
        "crossorigin": "anonymous",
        "fallback": "/pro/static/static/js/app.client.8dc14fd2c9e52eef.js",
      },
    ],
    [
      "cdn://static/js/locale/en.3c6a5e7a55bb2ab4.js",
      {
        "async": "",
        "crossorigin": "anonymous",
        "fallback": "/pro/static/static/js/locale/en.3c6a5e7a55bb2ab4.js",
      },
    ],
    [
      "cdn://static/js/vendors.0e456297851f0a3a.js",
      {
        "async": "",
        "crossorigin": "anonymous",
        "fallback": "/pro/static/static/js/vendors.0e456297851f0a3a.js",
      },
    ],
  ],
  "styles": [],
}
`;

exports[`resourcesUtils > prepareDefaultResources > should prepare default sources structure from provided manifest file 1`] = `
{
  "esScripts": [
    [
      "/static/js.es/app.client.2106a34c6b8bbad8.js",
      {
        "async": "",
        "crossorigin": "anonymous",
      },
    ],
    [
      "/static/js.es/locale/en.371127fdbfbe93d2.js",
      {
        "async": "",
        "crossorigin": "anonymous",
      },
    ],
    [
      "/static/js.es/vendors.a873907f25297544.js",
      {
        "async": "",
        "crossorigin": "anonymous",
      },
    ],
  ],
  "esStyles": [
    [
      "/static/css/app.d0ad44d05f82db5f.css",
      {
        "preload": true,
        "rel": "stylesheet",
      },
    ],
  ],
  "scripts": [
    [
      "/static/js/app.client.8dc14fd2c9e52eef.js",
      {
        "async": "",
        "crossorigin": "anonymous",
      },
    ],
    [
      "/static/js/locale/en.3c6a5e7a55bb2ab4.js",
      {
        "async": "",
        "crossorigin": "anonymous",
      },
    ],
    [
      "/static/js/vendors.0e456297851f0a3a.js",
      {
        "async": "",
        "crossorigin": "anonymous",
      },
    ],
  ],
  "styles": [],
}
`;

exports[`resourcesUtils > prepareDefaultResources > should skip compilations without assets 1`] = `
{
  "esScripts": [
    [
      "/static/js.es/app.client.2106a34c6b8bbad8.js",
      {
        "async": "",
        "crossorigin": "anonymous",
      },
    ],
    [
      "/static/js.es/locale/en.371127fdbfbe93d2.js",
      {
        "async": "",
        "crossorigin": "anonymous",
      },
    ],
    [
      "/static/js.es/vendors.a873907f25297544.js",
      {
        "async": "",
        "crossorigin": "anonymous",
      },
    ],
  ],
  "esStyles": [
    [
      "/static/css/app.d0ad44d05f82db5f.css",
      {
        "preload": true,
        "rel": "stylesheet",
      },
    ],
  ],
  "scripts": [],
  "styles": [],
}
`;

exports[`resourcesUtils > renderScript > should wrap script contents in script tag 1`] = `"<script id="ima-revival-settings">function () {}</script>"`;

exports[`resourcesUtils prepareDefaultResources should add fallbacks when IMA_PUBLIC_PATH is defined 1`] = `
{
  "esScripts": [
    [
      "/static/js.es/app.client.2106a34c6b8bbad8.js",
      {
        "async": "",
        "crossorigin": "anonymous",
      },
    ],
    [
      "/static/js.es/locale/en.371127fdbfbe93d2.js",
      {
        "async": "",
        "crossorigin": "anonymous",
      },
    ],
    [
      "/static/js.es/vendors.a873907f25297544.js",
      {
        "async": "",
        "crossorigin": "anonymous",
      },
    ],
  ],
  "esStyles": [
    [
      "/static/css/app.d0ad44d05f82db5f.css",
      {
        "preload": true,
        "rel": "stylesheet",
      },
    ],
  ],
  "scripts": [
    [
      "/static/js/app.client.8dc14fd2c9e52eef.js",
      {
        "async": "",
        "crossorigin": "anonymous",
      },
    ],
    [
      "/static/js/locale/en.3c6a5e7a55bb2ab4.js",
      {
        "async": "",
        "crossorigin": "anonymous",
      },
    ],
    [
      "/static/js/vendors.0e456297851f0a3a.js",
      {
        "async": "",
        "crossorigin": "anonymous",
      },
    ],
  ],
  "styles": [],
}
`;

exports[`resourcesUtils prepareDefaultResources should add fallbacks when IMA_PUBLIC_PATH is defined, with proper custom config publicPath 1`] = `
{
  "esScripts": [
    [
      "cdn://static/js.es/app.client.2106a34c6b8bbad8.js",
      {
        "async": "",
        "crossorigin": "anonymous",
        "fallback": "/pro/static/static/js.es/app.client.2106a34c6b8bbad8.js",
      },
    ],
    [
      "cdn://static/js.es/locale/en.371127fdbfbe93d2.js",
      {
        "async": "",
        "crossorigin": "anonymous",
        "fallback": "/pro/static/static/js.es/locale/en.371127fdbfbe93d2.js",
      },
    ],
    [
      "cdn://static/js.es/vendors.a873907f25297544.js",
      {
        "async": "",
        "crossorigin": "anonymous",
        "fallback": "/pro/static/static/js.es/vendors.a873907f25297544.js",
      },
    ],
  ],
  "esStyles": [
    [
      "cdn://static/css/app.d0ad44d05f82db5f.css",
      {
        "fallback": "/pro/static/static/css/app.d0ad44d05f82db5f.css",
        "preload": true,
        "rel": "stylesheet",
      },
    ],
  ],
  "scripts": [
    [
      "cdn://static/js/app.client.8dc14fd2c9e52eef.js",
      {
        "async": "",
        "crossorigin": "anonymous",
        "fallback": "/pro/static/static/js/app.client.8dc14fd2c9e52eef.js",
      },
    ],
    [
      "cdn://static/js/locale/en.3c6a5e7a55bb2ab4.js",
      {
        "async": "",
        "crossorigin": "anonymous",
        "fallback": "/pro/static/static/js/locale/en.3c6a5e7a55bb2ab4.js",
      },
    ],
    [
      "cdn://static/js/vendors.0e456297851f0a3a.js",
      {
        "async": "",
        "crossorigin": "anonymous",
        "fallback": "/pro/static/static/js/vendors.0e456297851f0a3a.js",
      },
    ],
  ],
  "styles": [],
}
`;

exports[`resourcesUtils prepareDefaultResources should prepare default sources structure from provided manifest file 1`] = `
{
  "esScripts": [
    [
      "/static/js.es/app.client.2106a34c6b8bbad8.js",
      {
        "async": "",
        "crossorigin": "anonymous",
      },
    ],
    [
      "/static/js.es/locale/en.371127fdbfbe93d2.js",
      {
        "async": "",
        "crossorigin": "anonymous",
      },
    ],
    [
      "/static/js.es/vendors.a873907f25297544.js",
      {
        "async": "",
        "crossorigin": "anonymous",
      },
    ],
  ],
  "esStyles": [
    [
      "/static/css/app.d0ad44d05f82db5f.css",
      {
        "preload": true,
        "rel": "stylesheet",
      },
    ],
  ],
  "scripts": [
    [
      "/static/js/app.client.8dc14fd2c9e52eef.js",
      {
        "async": "",
        "crossorigin": "anonymous",
      },
    ],
    [
      "/static/js/locale/en.3c6a5e7a55bb2ab4.js",
      {
        "async": "",
        "crossorigin": "anonymous",
      },
    ],
    [
      "/static/js/vendors.0e456297851f0a3a.js",
      {
        "async": "",
        "crossorigin": "anonymous",
      },
    ],
  ],
  "styles": [],
}
`;

exports[`resourcesUtils prepareDefaultResources should skip compilations without assets 1`] = `
{
  "esScripts": [
    [
      "/static/js.es/app.client.2106a34c6b8bbad8.js",
      {
        "async": "",
        "crossorigin": "anonymous",
      },
    ],
    [
      "/static/js.es/locale/en.371127fdbfbe93d2.js",
      {
        "async": "",
        "crossorigin": "anonymous",
      },
    ],
    [
      "/static/js.es/vendors.a873907f25297544.js",
      {
        "async": "",
        "crossorigin": "anonymous",
      },
    ],
  ],
  "esStyles": [
    [
      "/static/css/app.d0ad44d05f82db5f.css",
      {
        "preload": true,
        "rel": "stylesheet",
      },
    ],
  ],
  "scripts": [],
  "styles": [],
}
`;

exports[`resourcesUtils renderScript should wrap script contents in script tag 1`] = `"<script id="ima-revival-settings">function () {}</script>"`;
